README.md
![logo](./static/cyberlux-logo.jpg?raw=true "CYBERSECURITY Luxembourg")

<table>
<tr>
  <td>API checks</td>
</tr>
<tr>
  <td>Container builder</td>
  <td><a href="https://github.com/CybersecurityLuxembourg/openxeco/actions/workflows/oxe-api_docker.yml"><img src="https://github.com/CybersecurityLuxembourg/openxeco/actions/workflows/oxe-api_docker.yml/badge.svg" /></a></td>
</tr>
<tr>
  <td>Bandit Workflow</td>
  <td><a href="https://github.com/CybersecurityLuxembourg/openxeco/actions/workflows/oxe-api_pycqa-bandit.yml"><img src="https://github.com/CybersecurityLuxembourg/openxeco/actions/workflows/oxe-api_pycqa-bandit.yml/badge.svg" /></a></td>
</tr>
<tr>
  <td>Prospector Workflow</td>
  <td><a href="https://github.com/CybersecurityLuxembourg/openxeco/actions/workflows/oxe-api_pycqa-prospector.yml"><img src="https://github.com/CybersecurityLuxembourg/openxeco/actions/workflows/oxe-api_pycqa-prospector.yml/badge.svg" /></a></td>
</tr>
</table>

<table>
<tr>
  <td>Admin web</td>
</tr>
<tr>
  <td>Package builder</td>
  <td><a href="https://github.com/CybersecurityLuxembourg/openxeco/actions/workflows/oxe-web-admin_package.yml"><img src="https://github.com/CybersecurityLuxembourg/openxeco/actions/workflows/oxe-web-admin_package.yml/badge.svg" /></a></td>
</tr>
</table>

<table>
<tr>
  <td>Community web</td>
</tr>
<tr>
  <td>Package builder</td>
  <td><a href="https://github.com/CybersecurityLuxembourg/openxeco/actions/workflows/oxe-web-community_package.yml"><img src="https://github.com/CybersecurityLuxembourg/openxeco/actions/workflows/oxe-web-community_package.yml/badge.svg" /></a></td>
</tr>
</table>

<table>
<tr>
  <td>Licence</td>
</tr>
<tr>
  <td>FOSSA</td>
  <td><a href="https://app.fossa.com/projects/git%2Bgithub.com%2FCybersecurityLuxembourg%2Fopenxeco-core?ref=badge_shield" alt="FOSSA Status"><img src="https://app.fossa.com/api/projects/git%2Bgithub.com%2FCybersecurityLuxembourg%2Fopenxeco-core.svg?type=shield"/></a></td>
</tr>
</table>

<table>
<tr>
  <td>Social networks</td>
</tr>
<tr>
  <td>Twitter</td>
  <td><a href="https://twitter.com/cyberluxembourg"><img src="https://img.shields.io/twitter/follow/cyberluxembourg.svg?style=social&label=Follow" /></a></td>
</tr>
</table>

# Set up an instance

## For development

To set up the dev environment, please see those sub-project README files:

- [oxe-api/README.md](oxe-api/README.md)
- [oxe-web-admin/README.md](oxe-web-admin/README.md)
- [oxe-web-community/README.md](oxe-web-community/README.md)

## For testers

If you want to set up a local instance to test the project, please follow these instructions:

### Install docker

[Get Docker](https://docs.docker.com/get-docker/)

Linux:

```
$ sudo mkdir -p /etc/apt/keyrings/
$ curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo gpg --dearmor -o /etc/apt/keyrings/docker.gpg
$ echo "deb [arch=$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.gpg] https://download.docker.com/linux/ubuntu \
  $(lsb_release -cs) stable" | sudo tee /etc/apt/sources.list.d/docker.list > /dev/null
$ sudo apt update && sudo apt install docker-ce docker-ce-cli containerd.io docker-compose-plugin
$ sudo adduser <your-oxe-user> docker
$ newgrp docker
# If you want to verify your docker install run: docker run hello-world
```

### Install and run the openXeco containers and its dependencies

```
$ docker network create openxeco
$ docker run -d \
    --network openxeco \
    --network-alias mariadb \
    -p 3306:3306 \
    -e MARIADB_ROOT_PASSWORD=E4syPass \
    mariadb:10.7.3
$ docker run -d \
  --network openxeco \
  --network-alias smtp \
  -p 1025:1025 \
  -p 1080:1080 \
  reachfive/fake-smtp-server
$ docker build \
    -f openxeco-core-oxe-web-admin/Dockerfile \
    -t oxe-web-admin \
    --build-arg TARGET_DIR=openxeco-core-oxe-web-admin \
    https://github.com/CybersecurityLuxembourg/openxeco-core/releases/latest/download/openxeco-core-oxe-web-admin.tar.gz
$ docker run -d -p 3000:80 oxe-web-admin
$ docker build \
    -f openxeco-core-oxe-web-community/Dockerfile \
    -t oxe-web-community \
    --build-arg TARGET_DIR=openxeco-core-oxe-web-community \
    https://github.com/CybersecurityLuxembourg/openxeco-core/releases/latest/download/openxeco-core-oxe-web-community.tar.gz
$ docker run -d -p 3001:80 oxe-web-community
$ docker run -d -p 5000:5000 \
    --network openxeco \
    -e ENVIRONMENT=dev \
    -e JWT_SECRET_KEY=my_secret_developer_key \
    -e DB_HOSTNAME=mariadb \
    -e DB_PORT=3306 \
    -e DB_NAME=OPENXECO \
    -e DB_USERNAME=root \
    -e DB_PASSWORD=E4syPass \
    -e MAIL_SERVER=smtp \
    -e MAIL_PORT=1025 \
    -e MAIL_USE_TLS=True \
    -e MAIL_USE_SSL=False \
    -e MAIL_DEFAULT_SENDER=my-default-sender@example.org \
    -e IMAGE_FOLDER=/image_folder \
    -e DOCUMENT_FOLDER=/document_folder \
    -e INITIAL_ADMIN_EMAIL=my-default-admin@example.org \
    ghcr.io/cybersecurityluxembourg/openxeco-core-oxe-api:latest
```

### Install the app for dev with use docker-compose

- Run `docker-compose up`. Optionally can be used `-d` option to run as daemon.


### Enjoy the solution

Access the administrator interface:
- http://localhost:3000

Access the community interface:
- http://localhost:3001

Access the API documentation:
- http://localhost:5000
<br /><small>Note: if you are using Mac maybe port 5000 will be already in use. (check [solution](https://medium.com/pythonistas/port-5000-already-in-use-macos-monterey-issue-d86b02edd36c))</small>

Access the emails sent on the SMTP mock:
- http://localhost:1080

An initial account is created with the following email: "my-default-admin@example.org"
And a default password: "password"

## For production server

To set up the production instance, please see this file:

- [doc/INSTALL_SERVER.md](doc/INSTALL_SERVER.md)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

## [1.17.0] - 2023-10-25

### Changed

- API resource public/get_public_object_count: add filters by object type on taxonomy category
- Remove docker network as "external" in the general docker compose configuration
- Zoom the entity map on Europe instead of Luxembourg on oxe-web-admin
- Database structure: make the "FormQuestion.reference" from 20 to 30 chars long
- Database structure: make the "FormQuestion.options" from 500 chars long to TEXT
- Fix docker image version for API

### Added

- Forgotten "Accept privacy policy" field on user item on oxe-web-admin

### Fixed

- Campaign template verification on the "Verify and send" tab on oxe-web-admin
- Date fields that is fronting the clicks on select field options

## [1.16.9] - 2023-08-22

### Added

- BASE_DOMAIN environment variable for oxe-api

### Fixed

- Patch from v1.16.8 that is not working when using reverse proxy

## [1.16.8] - 2023-08-22

### Fixed

- Vulnerability: untrustworthy link injection in automatic email

## [1.16.7] - 2023-08-16

### Fixed

- Fix marker display on address geolocation selector on oxe-web-admin

## [1.16.6] - 2023-08-10

### Changed

- Display tab menu with ellipsis and text display on hover on oxe-web-admin and oxe-web-community
- Update package-lock.json on oxe-web-admin
- Update package-lock.json on oxe-web-community
- Update library version requirements on oxe-api

### Fixed

- Treat dependabot pull requests
- Typo

## [1.16.5] - 2023-08-08

### Changed

- API resource: private/update_my_form_answer to consider last_date column
- API resource: form/update_form_answer to consider last_date column

### Added

- last_date column on FormAnswer table
- last change date and system date on FormAnswer item on oxe-web-admin
- last change date and system date on FormAnswer item on oxe-web-admin
- MAIL_REPLY_TO environment variable for oxe-api

## [1.16.4] - 2023-07-27

### Fixed

- User section on FormAnswer item on owe-web-admin
- Answer deletion when the user provides empty answer on oxe-web-community
- Typo on oxe-web-community

### Changed

- Minor UI modifications on Entity logo tab on oxe-web-community

## [1.16.3] - 2023-07-12

### Fixed

- Request item display when no user ID on oxe-web-admin

## [1.16.2] - 2023-07-12

### Fixed

- Entity counting per taxonomy value on public/get_public_object_count resource

## [1.16.1] - 2023-07-11

### Fixed

- Package building for oxe-web-admin

### Changed

- Updated the package-lock.json for oxe-web-admin and oxe-web-community

## [1.16.0] - 2023-07-11

### Changed

- Review UI on Contact page of oxe-web-community
- Review UI on Form page of oxe-web-community
- Color style for Tab component on oxe-web-admin and oxe-web-community
- Improve color management in styling of oxe-web-community
- Review UI of logo management on oxe-web-admin
- Move the "Scheduled tasks" to "Tasks" page as "Launch scan" on oxe-web-admin

### Added

- "url" field type on oxe-web-admin and oxe-web-community
- Control on URL fields on oxe-web-admin and oxe-web-community
- DialogHint component on oxe-web-admin
- DialogError on oxe-web-admin and oxe-web-community
- Default logo and favicon
- API resource: setting/delete_favicon
- API resource: setting/delete_logo
- API resource: public/add_public_request - in order to handle public contact forms
- API resource: public/get_public_object_count
- Documentation on "Settings" page on oxe-web-admin
- Documentation on "Profile" page for notification system on oxe-web-admin
- Email notification to admins on new community article (private/add_article)

### Fixed

- Typo and text on documentation of owe-web-community
- Contact message creation on oxe-web-community by correcting database structure via migration
- Allow modification of user email via oxe-web-admin
- SQL querying requiring GROUP BY with the dockerized madiadb
- Show all optional field for article dialog, whatever the article type on oxe-web-admin

## [1.15.5] - 2023-04-04

### Fixed

- Dockerized API init

## [1.15.4] - 2023-03-30

### Fixed

- Fix entity extraction API resource

## [1.15.3] - 2023-03-09

### Changed

- "cryptography" package version for oxe-api

### Added

- Documentation of the configuration of the API

### Fixed

- Contact message creation on oxe-web-community

## [1.15.2] - 2023-02-15

### Fixed

- Cookie for password reseting on server instance on oxe-web-admin and oxe-web-community

## [1.15.1] - 2023-02-15

### Changed

- Documentation on the main README for the developers

### Fixed

- SMTP credentials for the development environment

## [1.15.0] - 2023-02-14

### Changed

- Persistence of email templates in the database
- user/update_user API resource: allow email modification
- public/get_public_document API resource: Add possibility to retrieve document by its ID
- user/update_user API resource: restrict field modification to admins
- UI on contact details of contact page of oxe-web-community
- UI/UX on "global" section of setting page of oxe-web-admin
- rename API module utils.re to utils.regex
- Require a "fresh" access token for all resources but account/unsubscribe

### Added

- Email notification to administrators on new request
- Default template for new request notification email
- Docker compose for development environment
- Database: accept_request_notification columns on User table
- Database: accept_terms_and_conditions columns on User table
- T&C/PP settings on oxe-web-admin for oxe-web-community
- T&C/PP dialog when not accepted on oxe-web-community
- Delete account button oxe-web-community
- GitHub url field for entites
- Mastodon url field for entities
- public/get_public_articles resource: add IDs filtering parameter
- account/unsubscribe resource
- campaign/send_campaign_draft resource
- mail/get_mail_addresses: add consider_communication_acceptance param
- Closing date for entity on DB, API resources, oxe-web-admin and oxe-web-community

config/config.py
import os

ENVIRONMENT = os.getenv('ENVIRONMENT', 'dev')

JWT_SECRET_KEY = os.getenv('JWT_SECRET_KEY', "my_secret_developer_key")

DB_CONFIG = {
    'drivername': 'mysql+pymysql',
    'username': os.getenv('DB_USERNAME', 'root'),
    'password': os.getenv('DB_PASSWORD', 'DBDevSMILE20'),
    'host': os.getenv('DB_HOSTNAME', 'localhost'),
    'port': os.getenv('DB_PORT', '3306'),
    'database': os.getenv('DB_NAME', 'CYBERLUX'),
}

MAIL_SERVER = os.getenv('MAIL_SERVER', "localhost")
MAIL_PORT = os.getenv('MAIL_PORT', "1025")
MAIL_USERNAME = os.getenv('MAIL_USERNAME', "cyberlux-app@securitymadein.lu")
MAIL_PASSWORD = os.getenv('MAIL_PASSWORD', None)
MAIL_USE_TLS = os.getenv('MAIL_USE_TLS', "False")
MAIL_USE_SSL = os.getenv('MAIL_USE_SSL', "False")
MAIL_DEFAULT_SENDER = os.getenv('MAIL_DEFAULT_SENDER', "cyberlux-app@securitymadein.lu")

IMAGE_FOLDER = os.getenv('IMAGE_FOLDER', "C:/image_folder")

JOBADDER_CLIENT_ID = "See JobAdder account"
JOBADDER_CLIENT_SECRET = "See JobAdder account"
JOBADDER_REFRESH_TOKEN = os.getenv('JOBADDER_REFRESH_TOKEN', None)
JOBADDER_ACCESS_TOKEN = os.getenv('JOBADDER_ACCESS_TOKEN', None)

FRONTEND_URL = {
    'dev': ['http://localhost:3000'],
    'test': ['https://test-db.cy.lu'],
    'production': ['https://db.cy.lu']
}[ENVIRONMENT]

PUBLIC_FRONTEND_URL = {
    'dev': ['http://localhost:3002'],
    'test': ['https://test-secin.cy.lu'],
    'production': ['https://cybersecurity-luxembourg.com', 'https://cy.lu'],
}[ENVIRONMENT]

PUBLIC_ECOSYSTEM_FRONTEND_URL = {
    'dev': ['http://localhost:3003'],
    'test': ['https://test-eco.cy.lu'],
    'production': ['https://ecosystem.cybersecurity-luxembourg.com', 'https://eco.cy.lu'],
}[ENVIRONMENT]
